src/adc.rs

```diff
@@ -449,6 +449,244 @@ impl_adc!(ADC1: adc_unit_t_ADC_UNIT_1);
 #[cfg(not(any(esp32c2, esp32h2, esp32c5, esp32c6, esp32p4)))] // TODO: Check for esp32c5 and esp32p4
 impl_adc!(ADC2: adc_unit_t_ADC_UNIT_2);
 
+#[cfg(all(not(feature = "riscv-ulp-hal"), not(esp_idf_version_major = "4")))]
+pub mod oneshot {
+    use core::marker::PhantomData;
+
+    use esp_idf_sys::*;
+
+    use crate::gpio::ADCPin;
+    use crate::peripheral::Peripheral;
+    use crate::peripheral::PeripheralRef;
+
+    use super::attenuation::adc_atten_t;
+    use super::config::Resolution;
+    use super::Adc;
+
+    pub mod config {
+        use super::adc_atten_t;
+        use super::Resolution;
+
+        #[derive(Debug, Copy, Clone, Default)]
+        pub struct AdcChannelConfig {
+            pub attenuation: adc_atten_t,
+            pub resolution: Resolution,
+            #[cfg(any(esp_idf_comp_esp_adc_cal_enabled, esp_idf_comp_esp_adc_enabled))]
+            pub calibration: bool,
+        }
+    }
+
+    pub struct AdcChannelDriver<'d, T: ADCPin> {
+        pin: PeripheralRef<'d, T>,
+        calibration: Option<adc_cali_handle_t>,
+    }
+
+    impl<'d, T: ADCPin> AdcChannelDriver<'d, T> {
+        pub fn new(
+            adc: &AdcDriver<'d>,
+            pin: impl Peripheral<P = T> + 'd,
+            config: config::AdcChannelConfig,
+        ) -> Result<AdcChannelDriver<'d, T>, EspError> {
```

Comment:
Nit: `Self` instead of `AdcChannelDriver<'d, T>`, and yes - I have to fix this in the 4.X oneshot driver as well.